Lifesaver necklace

TsunamiAnt opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

Adds the lifesaver necklace to the mining voucher as well as the mining vendor.
It's a small matchbook sized plastitanium box on a string that screams and whines and calls over radio and makes boopy noises when it's user dies. It also shows up on GPS.

### It only works on lavaland.

## Why It's Good For The Game

**Never complain to me about being forgotten again.**

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.
-->

## Testing Photographs and Procedure
<!-- Include any screenshots/videos/debugging steps of the modified code functioning successfully, ideally including edge cases. -->
<details>
<summary>Screenshots&Videos</summary>

<img width="694" height="696" alt="image" src="https://github.com/user-attachments/assets/95967240-2071-4119-b83c-2a9655230d25" />

![dreamseeker_095o6vElMx](https://github.com/user-attachments/assets/a33cf87b-ed33-4d50-8e4a-5dc621a13e31)

![dreamseeker_0qUPau3sJg](https://github.com/user-attachments/assets/393df72c-dcca-4e0a-84f2-bfa2754aaad1)

<img width="866" height="95" alt="image" src="https://github.com/user-attachments/assets/35559a29-d123-42cb-9c75-9ef3b5dffc1d" />

<img width="617" height="223" alt="image" src="https://github.com/user-attachments/assets/9617e375-2224-400a-9537-1e527d01dddb" />

<img width="711" height="36" alt="image" src="https://github.com/user-attachments/assets/57394733-04c4-4967-bacc-143f323d84b2" />

</details>

## Changelog
:cl:
add: Added Lifesaver necklace, a death-notifier for miners.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

TsunamiAnt committed:
Adds it

mrmanlikesbt commented on code/modules/mining/equipment/lifesaver.dm:
```diff
@@ -0,0 +1,79 @@
+/obj/item/clothing/neck/necklace/lifesaver/attack_self(mob/user)
+	if(!active)
+		return
+	to_chat(user, "You press your finger on the touch-button, disarming the necklace's alert state.")
```
use some span here. `span_notice` maybe?

TsunamiAnt replied:
oof yeah right definitely, i knew it looked weird

mrmanlikesbt commented on code/modules/mining/equipment/lifesaver.dm:
```diff
@@ -0,0 +1,79 @@
+	to_chat(user, "You press your finger on the touch-button, disarming the necklace's alert state.")
+	disableAlert()
+
+/obj/item/clothing/neck/necklace/lifesaver/proc/enableAlert()
```
no using camelcase lil bro
```suggestion
/obj/item/clothing/neck/necklace/lifesaver/proc/enable_alert()
```

mrmanlikesbt commented on code/modules/mining/equipment/lifesaver.dm:
```diff
@@ -0,0 +1,79 @@
+	AddComponent(/datum/component/gps, "#ACTIVE LIFESAVER - RESCUE REQUIRED#")
+	addtimer(CALLBACK(src, PROC_REF(alertRoutine)), 5 SECONDS)
+
+/obj/item/clothing/neck/necklace/lifesaver/proc/disableAlert()
```
no using camelcase lil bro
```suggestion
/obj/item/clothing/neck/necklace/lifesaver/proc/disable_alert()
```

mrmanlikesbt commented on code/modules/mining/equipment/lifesaver.dm:
```diff
@@ -0,0 +1,79 @@
+	active = FALSE
+
+//Loops every few seconds
+/obj/item/clothing/neck/necklace/lifesaver/proc/alertRoutine()
```
no using camelcase lil bro
```suggestion
/obj/item/clothing/neck/necklace/lifesaver/proc/alert_routine()
```

mrmanlikesbt commented on code/modules/mining/equipment/lifesaver.dm:
```diff
@@ -0,0 +1,79 @@
+	icon_state = "lifesaver"
+	worn_icon_state = "lifesaver"
+
+	qdel(src.GetComponent(/datum/component/gps))
```
```suggestion
	qdel(GetComponent(/datum/component/gps))
```

mrmanlikesbt commented on code/modules/mining/equipment/lifesaver.dm:
```diff
@@ -0,0 +1,79 @@
+	worn_icon_state = "lifesaver_active"
+	active_owner.regenerate_icons()
+
+	radio.talk_into(src, "ALERT - RESCUE REQUIRED")
```
maybe?

```suggestion
	radio.talk_into(src, "Alert - Rescue required at [get_area(src)]!")
```

mrmanlikesbt replied:
Resolved on discord

mrmanlikesbt commented on code/modules/mining/equipment/lifesaver.dm:
```diff
@@ -0,0 +1,79 @@
+		RegisterSignal(active_owner, COMSIG_MOB_DEATH, PROC_REF(enableAlert))
+		return
+	else if(slot != ITEM_SLOT_POCKETS && slot != ITEM_SLOT_BACKPACK)
+		say("Biomonitor error.")
```
This is going to be annoying getting a message every time you put it into your backpack/pockets/whatever

```suggestion
```

TsunamiAnt replied:
ah no this actually prevents that message from appearing in those cases.

mrmanlikesbt replied:
Oh my bad. Even still, why have this message if the only possible slot you can equip it to is your neck (afaik)?

TsunamiAnt replied:
It kinda felt cool to pick it up and have it immediately recognize something is happening. Iunno vibe stuff.

If you want i can remove it?

mrmanlikesbt replied:
Ya

TsunamiAnt replied:
done!

mrmanlikesbt commented on code/modules/mining/equipment/lifesaver.dm:
```diff
@@ -0,0 +1,79 @@
+/obj/item/clothing/neck/necklace/lifesaver/Initialize(mapload)
+	. = ..()
+
+	radio = new/obj/item/radio(src)
```
```suggestion
	radio = new /obj/item/radio(src)
```

mrmanlikesbt commented on code/modules/mining/equipment/lifesaver.dm:
```diff
@@ -0,0 +1,79 @@
+	var/mob/living/carbon/human/active_owner
+	var/active = FALSE
+
+	var/radioCounter
```
no using camelcase lil bro
```suggestion
	var/radio_counter
```

mrmanlikesbt commented on code/modules/mining/equipment/lifesaver.dm:
```diff
@@ -0,0 +1,79 @@
+/obj/item/clothing/neck/necklace/lifesaver
+	name = "Acrux-brand life-saving necklace"
+	desc = "You won't be getting any style points with this boring plastitanium box around your neck, but at least you will get rescued in time after it detects your untimely demise and begins causing a massive ruckus over it."
+	icon = 'icons/obj/clothing/neck.dmi'
```
already defined in `/obj/item/clothing/neck/necklace`

```suggestion
```

mrmanlikesbt commented on code/modules/mining/equipment/lifesaver.dm:
```diff
@@ -0,0 +1,79 @@
+//Loops every few seconds
+/obj/item/clothing/neck/necklace/lifesaver/proc/alertRoutine()
+	if(active)
+
```
```suggestion
```

mrmanlikesbt commented on code/modules/mining/equipment/lifesaver.dm:
```diff
@@ -0,0 +1,79 @@
+
+		// Periodic unhelpful radio announcements
+		if(radioCounter >= 20)
+			radio.talk_into(src, "ALERT - RESCUE REQUIRED")
```
maybe?

```suggestion
			radio.talk_into(src, "Alert - Rescue required at [get_area(src)]!")
```

mrmanlikesbt replied:
Resolved on discord

mrmanlikesbt commented on code/modules/mining/equipment/lifesaver.dm:
```diff
@@ -0,0 +1,79 @@
+			say("ALERT - RESCUE REQUIRED")
+			radioCounter = 0
+		else
+			radioCounter += 1
```
```suggestion
			radioCounter++
```

mrmanlikesbt reviewed (dismissed):
Attribute the source of livesaver.ogg

TsunamiAnt committed:
Mrman's demands

TsunamiAnt committed:
does this fix it?

TsunamiAnt committed:
We do actually want to overwrite it, oops

TsunamiAnt committed:
Bacon wants it to only work on lavaland

TsunamiAnt committed:
remove locs

TsunamiAnt committed:
remove all the vibe

TsunamiAnt committed:
bugfix

TsunamiAnt committed:
aye

Rukofamicom reviewed (dismissed):
Give it a 1 minute delay instead of 15 seconds and have it repeat every minute 
Make them available in medbay vendors (4 stock?) + 1 in each mining locker 
Make them work on station too, as a sort of enhanced sensors

Upping the delay to a full minute gives ample time for antagonists encountering someone that is wearing one to remove it or get the fuck away from their victim. 

Acting in discord like this PR puts it entirely as the miner's responsibility to be found really soured my view of this PR when you have chosen to make this item the most expensive item in the vendor, when the average miner doesn't make more than 20,000 points in a round. The station should already be checking on miners and it's ridiculous to act like it's entirely the miner's responsibility to want to be found beyond taking a GPS on top of making it so expensive.

TsunamiAnt committed:
Adjusts lifesaver beacon functionality.

Increases the lifesaver activation delay and introduces a shutoff mechanism for radio announcements after a certain period, while still keeping the GPS active.
This prevents constant radio spam while still allowing for location tracking.

TsunamiAnt committed:
makes them more available

PowerfulBacon commented on code/modules/mining/equipment/lifesaver.dm:
```diff
@@ -0,0 +1,95 @@
+	QDEL_NULL(radio)
+	return ..()
+
+/obj/item/clothing/neck/necklace/lifesaver/equipped(mob/user, slot)
```
This does not account for being dropped directly without going through the hands first.

PowerfulBacon replied:
When unequipped or dropped, it should disable its pinging
